chapter_hyperparameter_optimization/hyperband-intro.md

```diff
@@ -88,73 +104,93 @@ from d2l import tensorflow as d2l
+    return self.queue.pop()
 ```
 
-When we collected a new data point, we first update the searcher module. Afterwards we check if we already collect enough data points $n_i = \eta^{K - i}$ on the current rung level. If so, we sort all configurations based on their performance and estimate the top $\frac{1}{\eta}$ configuration and append them to the queue, which should be empty at this point.
+When we collected a new data point, we first update the searcher module. Afterwards
+we check if we already collect all data points on the current rung level. If so, we
+sort all configurations based on their performance and push the top $\frac{1}{\eta}$
+configurations into the queue.
 
 ```{.python .input  n=4}
 %%tab all
 @d2l.add_to_class(SuccessiveHalvingScheduler) #@save
 def update(self, config, error, info=None):
-    # determine the rung level r_i
-    ri = config['max_epochs']
-    # bookkeeping
-    self.observed_error_at_rungs[ri].append((config, error.cpu().numpy()))
+    ri = config['max_epochs']  # rung level r_i
     # update our searcher, e.g if we use Bayesian optimization later
     self.searcher.update(config, error, additional_info=info)     
-    # determine how many configurations should be evaluated on this rung level
-    ki = self.K - self.rung_levels.index(ri)
-    n = int(self.s * self.eta ** ki)
-    # if we observed all configuration on this rung level r_i, we estimate the top 1 / eta configuration and 
-    # add them to queue and promote them for the next rung level r_i+1
-    if len(self.observed_error_at_rungs[ri]) == n and ri < self.r_max:
-        best_performing_configurations = self.get_top_n_configurations(ri, n // self.eta)
-        for config in best_performing_configurations:
-            config['max_epochs'] = int(ri * self.eta)
-            self.queue.append(config)
+    if ri < r_max:
+        # bookkeeping
+        self.observed_error_at_rungs[ri].append((config, error.cpu().numpy()))
+        # determine how many configurations should be evaluated on this rung level
+        ki = self.K - self.rung_levels.index(ri)
+        ni = int(self.prefact * self.eta ** ki)
+        # if we observed all configuration on this rung level r_i, we estimate the
+        # top 1 / eta configuration, add them to queue and promote them for the
+        # next rung level r_{i+1}
+        if len(self.observed_error_at_rungs[ri]) >= ni:
+            kiplus1 = ki - 1
+            niplus1 = int(self.prefact * self.eta ** kiplus1)
+            best_performing_configurations = self.get_top_n_configurations(
+                rung_level=ri, n=niplus1
+            )
+            riplus1 = self.rung_levels[self.K - kiplus1]  # r_{i+1}
+            # queue may not be empty: insert new entries at the beginning
+            self.queue = [
+                dict(config, max_epochs=riplus1)
+                for config in best_performing_configurations
+            ] + self.queue
+            self.observed_error_at_rungs[ri] = []  # reset
```

Comment:
We should not conflate things. The complete data is probably already stored in the base classes, no?

This here needs to be reset, because there may be another round. It is also reset in your code, just at a different location.